gnucash/gnome-utils/gnc-file.c

```diff
@@ -1028,9 +1029,24 @@ gnc_post_file_open (GtkWindow *parent, const char * filename, gboolean is_readon
         g_free ( message );
     }
 
-    // Fix account color slots being set to 'Not Set', should run once on a book
     qof_event_suspend();
+
+    // Fix account color slots being set to 'Not Set', should run once on a book
     xaccAccountScrubColorNotSet (gnc_get_current_book());
+
+    /* Fix budget signs */
+    switch (scrub_budget_signs_check (new_book))
+    {
+    case 0:
+        break;
+    case 1:
+        scrub_budget_signs_add_feature (new_book);
+        break;
+    case 2:
+        if (gnc_verify_dialog (parent, FALSE, (_("A one-time fix should be applied to budget amounts so that they are stored internally with natural rather than reversed numbers. After this fix, please review budgets and ensure amounts are as expected. If this fix is not applied, the budget reports will be incorrect. Should the fix be applied? Please note after this fix, the data will be unreadable by GnuCash version 3.7 or earlier." ))))
+            scrub_budget_signs_doit (new_book);
+    }
```

Comment:
I'm still unsure we covered all possible scenarios:
* User applies fix, and accepts changes (the ideal case)
* User applies fix, rejects changes (back to square 1, will be asked again next time)
* User refuses fix, anything goes as we simply don't know what the user will do next. Will the user manually fix the budgets? Change sign reversal strategy? What? Assuming the user chose to fix the budgets manually, it would be annoying to get the same fix request again on next startup. And again and again until s/he finally allows the scrub to run once because that's currently the only way the feature flag gets set if there are already budgets. And perhaps that's exactly what the user didn't want to do! Perhaps our heuristics were wrong after all in a some obscure corner case situation.

So I think we're missing a solid path for the case where the user wants to fix the budgets manually.

Based on that I would propose a slightly different approach
First determine the state the current book is in:
* feature set or no budgets -> do nothing at file loading
* budgets but no fix needed -> inform the user we think budgets are in good shape and want to prevent older gnucash versions from messing this up. Explain that doing so will prevent 3.7 or older to open the book, and not doing so will mean the budgets may have sign issues when switching between 3.x and 4.x. Question will be repeated at book loading until user agrees to prevent loading in older versions or user makes a budget change.
* budgets that need fixing -> inform the user we think budgets need fixing and that doing so will prevent 3.7 or older from opening the book. Not doing so will continue to pop up this dialog when opening a book until the user agrees to fix or does the fixes him/herself. In both cases older versions will no longer be able to open the book.

Then we should also catch any budget value changes made directly by the user. On any such change check whether the feature flag is already set. If not pop up a warning explaining that making budget value changes will 
1. prevent older gnucash versions from opening the book and
2. stop gnucash from any future attempt to do natural sign budget  fixes in the future. So potentially the user may want to check any other budgets as well. If continue anyway, set feature flag and leave the rest to the user. If not revert attempted change.

Reply:
So, your view is:
* feature set, or no budgets -> do nothing
* budgets but no fix needed -> only applies if *all* budgets seem to follow 'reversed balance=none' which is extremely unlikely -- I think the Estimate tool only works when pref is credit-accounts
* budgets need fixing -> offer Yes/No dialog to "allow automatic fixing? if yes it'll be done silently. if no, you may need to fix budgets manually" which applies my fixes and sets budget.

And when budget cells are modified:
* if book is not featured, then show a pop up once: "please be advised editing budget manually will set a flag that will cause the datafile to be unreadable by 3.7 or earlier, please confirm you wish to continue editing?"  yes=set feature, no=abort.

This is not a bad idea. My original idea was that unfeatured datafiles' budgets were always fixed which mean we could remove some old code.

Reply:
> So, your view is:

It's almost my view...

> 
>     * feature set, or no budgets -> do nothing
> 
>     * budgets but no fix needed -> only applies if _all_ budgets seem to follow 'reversed balance=none' which is extremely unlikely -- I think the Estimate tool only works when pref is credit-accounts
> 
This is really `budgets but we *think* no fix needed`. Someone may be budgetting only reimbursed expenses for example, which would cause the heuristics to come to the wrong conclusion.

>     * budgets need fixing -> offer Yes/No dialog to "allow automatic fixing? if yes it'll be done silently. if no, you may need to fix budgets manually" which applies my fixes and sets budget.
> 
> 
> And when budget cells are modified:
> 
>     * if book is not featured, then show a pop up once: "please be advised editing budget manually will set a flag that will cause the datafile to be unreadable by 3.7 or earlier, please confirm you wish to continue editing?"  yes=set feature, no=abort.
> 
That's the idea though for the best user experience it should also test (again) whether the budget may require fixing and if we think so add an additional message that the signs may be wrong but manually editing the budget would prevent gnucash from offering to fix this on future runs.

> 
> This is not a bad idea. My original idea was that unfeatured datafiles' budgets were always fixed which mean we could remove some old code.

I think that would be a false assumption. Unfeatured datafile's budgets are the ones the may need fixing, because these have been created with older gnucash versions. Or did you mean something else ?

Reply:
> > * budgets but no fix needed -> only applies if _all_ budgets seem to follow 'reversed balance=none' which is extremely unlikely -- I think the Estimate tool only works when pref is credit-accounts
> 
> This is really `budgets but we *think* no fix needed`. Someone may be budgetting only reimbursed expenses for example, which would cause the heuristics to come to the wrong conclusion.

Ok -- in this case I am not sure I can reliably detect that budgets don't need fixing, i.e.  I think whereby sign-reversal will do nothing. Any ideas here?

> > And when budget cells are modified:
> > * if book is not featured, then show a pop up once: "please be advised editing budget manually will set a flag that will cause the datafile to be unreadable by 3.7 or earlier, please confirm you wish to continue editing?"  yes=set feature, no=abort.
> 
> That's the idea though for the best user experience it should also test (again) whether the budget may require fixing and if we think so add an additional message that the signs may be wrong but manually editing the budget would prevent gnucash from offering to fix this on future runs.

Ok I'll see what I can do.

> > This is not a bad idea. My original idea was that unfeatured datafiles' budgets were always fixed which mean we could remove some old code.
> 
> I think that would be a false assumption. Unfeatured datafile's budgets are the ones the may need fixing, because these have been created with older gnucash versions. Or did you mean something else ?

I meant: gnucash 4.x would aim to set the feature in every book (possibly after applying fixes)  which means that the old code to handle reversal could be removed. But it's fine to use your approach above.

Reply:
> > > * budgets but no fix needed -> only applies if _all_ budgets seem to follow 'reversed balance=none' which is extremely unlikely -- I think the Estimate tool only works when pref is credit-accounts
> > 
> > 
> > This is really `budgets but we *think* no fix needed`. Someone may be budgetting only reimbursed expenses for example, which would cause the heuristics to come to the wrong conclusion.
> 
> Ok -- in this case I am not sure I can reliably detect that budgets don't need fixing, i.e. I think whereby sign-reversal will do nothing. Any ideas here?

I agree there isn't a 100% reliable way of detecting this. Which is why I proposed to tell the user what we *think* is the right way forward and propose to act according to it:
- either we think a fix is needed, so propose to apply the fix, but offer the user to decide not to do so
- or we think budgets are correct, but we want to carve that in stone by by setting a flag. Again if the user is still using gnucash pre-3.8 on any other machines allow him to not set that flag until he's ready.

We might mention in these initial message already in case of decline the nagging will continue on future reopening of the book until either the user accepts the proposed solution or makes a manual change to any existing budget. End result will always be the book can only be opened in 3.8 or later from then on.

Having said all that, there's another way to look at this: pre-3.8 budget format is considered read-only in 4.x onward. So you could also consider to formulate the user messaging in that sense:

> Hi, we notice your budgets are still in 3.x format. 4.x has a new format to store budgets. We can convert automatically to this new format if you like. Beware that once converted you'll need at least gnucash 3.8 be able to open your book.
> On the other hand if you want to make changes to your budgets in 4.x you have to convert it to the new format. GnuCash 4.x can't write the old format any more.
> Note the conversion may involve a on-time sign correction on your budgeted values. Until you run the conversion some budgeted values may show with wrong signs. However for completeness there is a small chance the budgeted values show up with the wrong sign *after* the conversion. In that unfortunate case you'll have to manually correct these yourself.
> Would you like gnucash to convert your budgets now ?

This message focuses on the impact it has on the user:
- convert (with or without sign conversion which is now an implementation detail the user doesn't need to worry about), which has the following impact
  - budgets can be manipulated
  - books older than 3.8 can't open this book any more
  - have a small chance some budgeted values need manual sign fixes
- keep as is, in which case
  - budgets are read-only
  - several budgeted values are likely to be presented with wrong signs
  - but can still be manipulated in 3.7 and before

This also assumes my previous counter example is the most uncommon case and that your heuristics have a high chance of being right. I actually think they have.
In that case the decision to be able to open in pre-3.8 is for most users the more important factor.

Combined with this opening message, I would effectively make budgets read-only until the conversion has run. I believe we already test on the book's read-onlyness in budget views. It should not be too hard to test for a second read-only factor, being the natural sign feature being unset.

And as a convenience to users that skipped the conversion at book opening we can also add a menu item in the budget menu to convert budgets. That item would only be shown when the feature flag is not set.

What we lose with this scenario is the case where we think conversion is needed and the user believes it is not. In the previous scenario the user could just decline conversion, go to the budget and make one single change. Gnucash would then set the feature flag without any extra conversions. In the new scenario, the user eventually has to go through the conversion at some point. And if unlucky, that means some budgets would get the wrong signs. We have evaluated this is the uncommon case (a theoretical case I just maded up).

In the end I think the new scenario would be easier to cope with for the user. Asking the user to decide how to do the conversion may be stressful for all but the most experienced users.

My apologies for turning everything upside down once more...

Reply:
I think the current branch already functions pretty much as described. Only the string will need amend. So, please feel free to test this PR with a book with or without existing budgets, and try budget editor and reports.

> Combined with this opening message, I would effectively make budgets read-only until the conversion has run. I believe we already test on the book's read-onlyness in budget views. It should not be too hard to test for a second read-only factor, being the natural sign feature being unset.

fwiw the budget editor doesn't obey the readonly property. Thanks to earlier work in maint, both budget editor and reports can handle both unreversed and reversed.

Reply:
> And as a convenience to users that skipped the conversion at book opening we can also add a menu item in the budget menu to convert budgets. That item would only be shown when the feature flag is not set.

I think this is the only feature missing. I'd think this is overengineering...

Reply:
This whole line of reasoning is overthinking. We do this a lot: Fix a problem, set a feature. We usually don't even tell the user about it, but in this case the fix can misfire so we tell the user to check that it went OK.